package/src/version.rs

```diff
@@ -0,0 +1,250 @@
+//! This module contains everything to do with version numbers.
+
+use std::{num::ParseIntError, str::FromStr};
+
+use serde::{Deserialize, Serialize};
+
+/// A full semantic version, including prerelease and build metadata.
+#[serde_with::serde_as]
+#[derive(
+    Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, serde::Serialize, serde::Deserialize,
+)]
+pub struct FullSemVer {
+    pub major: u64,
+    pub minor: u64,
+    pub patch: u64,
+    #[serde(default = "String::default")]
+    pub pre: String,
+    #[serde(default = "String::default")]
+    pub build: String,
+}
+
+/// Our most-widely-used version type.
+///
+/// This drops the build metadata part (which we allow during parsing but
+/// ignore for all version-resolution purposes).
+///
+/// Possible optimizations:
+/// - shrink the numbers to `u32`
+/// - intern the prerelease tag. This needs to be done in a way that preserves
```

Comment:
Do you think it's worth the trouble, given that `pre` is probably rather small? Another solution is to use `SmolString` or the like.

Reply:
Probably not worth it, I just observed that the `semver` crate puts some effort into an optimized layout so I just wanted to record the possibility. Actually, I was more concerned about `pre` bloating the size of `SemVer` than about the size of the string: `Ident` is only 4 bytes, vs 24 for `String`.